### PR TITLE
Introduces custom excluded phrases

### PR DESCRIPTION
**Pull request by zonble**
The pull request uses a FastLM to load the phrases that that a user want to exclude from the phrase database. Now the BlockReadingBuilder reads data from three language models, the main input method language model, the user phrases, and excluded phrase. The builder looks up for Unigrams from the first and second models and use the last one as a filter.

A user can start to edit the file storing phrases to exclude from the input menu and then click on the "reload user phrases" menu item to make new excluded phrases to take effect. Actually I think it is an advanced function and there might be a Wiki page for it.

This solves requests like issue #123.

### CODE REVIEW

**Comment by zonble** [comment]
By the way, I moved the instance of these language models to another file. Since the main input controller is already a little bit too big, and the models are actually global but should not be a part of the input controller, I created a new "Language Model Manager" as the facade for all of the language models.

**Comment by zonble** [comment]
Temporarily close the PR since it is better to hide the additional language models to a new abstract language model.

File: Source/Engine/Gramambular/BlockReadingBuilder.h [hunk 1]
```diff
@@ -58,6 +58,8 @@ namespace Formosa {
             vector<string> readingsAtRange(size_t begin, size_t end) const;
 
             Grid& grid();
+
+            bool checkIfUnigramExistInVector(Unigram& unigram, vector<Unigram>vector);
```
Comment (lukhnos) [inline Source/Engine/Gramambular/BlockReadingBuilder.h]: No longer used. Remove?

File: Source/Engine/Gramambular/BlockReadingBuilder.h [hunk 2]
```diff
@@ -197,7 +197,17 @@ namespace Formosa {
         {
             return m_grid;
         }
-        
+
+        inline bool BlockReadingBuilder::checkIfUnigramExistInVector(Unigram& unigram, vector<Unigram>vector)
```
Comment (lukhnos) [inline Source/Engine/Gramambular/BlockReadingBuilder.h]: No longer used.

File: Source/Engine/McBopomofoLM.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+{
+    m_languageModel.close();
+    m_languageModel.open(languageModelDataPath);
+}
+
+void McBopomofoLM::loadUserPhrases(const char* userPhrasesDataPath,
+                                   const char* excludedPhrasesDataPath)
+{
+    m_userPhrases.close();
+    m_userPhrases.open(userPhrasesDataPath);
```
Comment (lukhnos) [inline Source/Engine/McBopomofoLM.cpp]: `LanguageModelManager` passes `""` to `userPhrasesDataPath` when creating Plain Bopomofo's LM. Perhaps allow `nullptr` here, and only call `open` if `userPhrasesDataPath` is not `nullptr`?

Same for `excludedPhrasesDataPath`.

I realize calling `::open("", O_RDONLY)` in `FastLM` will likely lead to a safely-guarded failure, but one more guard shouldn't hurt…

File: Source/Engine/McBopomofoLM.h [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                         const char* m_excludedPhrasesDataPath);
+
+    const vector<Bigram> bigramsForKeys(const string& preceedingKey, const string& key);
+    const vector<Unigram> unigramsForKey(const string& key);
+    bool hasUnigramsForKey(const string& key);
+
+protected:
+    FastLM m_languageModel;
+    FastLM m_userPhrases;
+    FastLM m_excluddePhrases;
```
Comment (lukhnos) [inline Source/Engine/McBopomofoLM.h]: Nit: `m_excludedPhrases`?

File: Source/LanguageModelManager.mm [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
++ (void)loadDataModels
+{
+    LTLoadLanguageModelFile(@"data", gLanguageModelMcBopomofo);
+    LTLoadLanguageModelFile(@"data-plain-bpmf", gLanguageModelPlainBopomofo);
+}
+
++ (void)loadUserPhrasesModel
+{
+    gLanguageModelMcBopomofo.loadUserPhrases([[self userPhrasesDataPathMcBopomofo] UTF8String], [[self excludedPhrasesDataPathMcBopomofo] UTF8String]);
+    gLanguageModelPlainBopomofo.loadUserPhrases("", [[self excludedPhrasesDataPathPlainBopomofo] UTF8String]);
```
Comment (lukhnos) [inline Source/LanguageModelManager.mm]: Let `loadUserPhrases` accept `nullptr`, and pass `nullptr` here?

**Review by lukhnos — APPROVED** [review]
LGTM.

**Pull request merged**